covenants/src/lib.rs

```diff
@@ -0,0 +1,1142 @@
+            self.tx_version.consensus_encode(&mut writer)?;
+            writer
+        };
+
+        // input specific values
+        let (previous_out, script_0, script_1, script_2, value, sequence) = {
+            let InputData {
+                previous_output,
+                script,
+                value,
+                sequence,
+            } = &self.input;
+
+            let third = script.len() / 3;
+
+            (
+                {
+                    let mut writer = Vec::new();
+                    previous_output.consensus_encode(&mut writer)?;
+                    writer
+                },
+                {
+                    let mut writer = Vec::new();
+                    script.consensus_encode(&mut writer)?;
+                    writer[..third].to_vec()
+                },
+                {
+                    let mut writer = Vec::new();
+                    script.consensus_encode(&mut writer)?;
+                    writer[third..2 * third].to_vec()
+                },
+                {
+                    let mut writer = Vec::new();
+                    script.consensus_encode(&mut writer)?;
+                    writer[2 * third..].to_vec()
+                },
+                {
+                    let mut writer = Vec::new();
+                    value.consensus_encode(&mut writer)?;
+                    writer
+                },
+                {
+                    let mut writer = Vec::new();
+                    sequence.consensus_encode(&mut writer)?;
+                    writer
+                },
+            )
+        };
+
+        // hashoutputs (only supporting SigHashType::All)
+        let other_outputs = {
+            let mut other_outputs = vec![];
+
+            if self.other_outputs.len() < 2 {
+                bail!("insufficient outputs");
+            }
+
+            if self.other_outputs.len() > 3 {
+                bail!("too many outputs");
+            }
+
+            for txout in self.other_outputs.iter() {
+                let mut output = Vec::new();
+                txout.consensus_encode(&mut output)?;
+
+                let middle = output.len() / 2;
+                other_outputs.push(output[..middle].to_vec());
+                other_outputs.push(output[middle..].to_vec());
+            }
+
+            // fill in space for missing principal change output
+            if other_outputs.len() == 4 {
+                other_outputs.push(vec![]);
+                other_outputs.push(vec![]);
+            }
+
+            other_outputs
+        };
+
+        let lock_time = {
+            let mut writer = Vec::new();
+            self.lock_time.consensus_encode(&mut writer)?;
+            writer
+        };
+
+        let sighash_type = {
+            let mut writer = Vec::new();
+            self.sighash_type.as_u32().consensus_encode(&mut writer)?;
+            writer
+        };
+
+        Ok(vec![
+            sig,
+            pk,
```

Comment:
@bonomat remind me: why do we need this pk in the witness stack? Don't we already have it in the script? Is it just for convenience?

Reply:
You are right, we have it already in the script `        .push_slice(&borrower_pk.serialize())
`. 
I don't think there was a particular reason for having it here again. IF we change it, we will need to make use of `OP_DUP` and move the `pk` in the script around a couple of times or add it to the script twice. (we need it for `op_checksig` and `op_checksigfromstack`).

Reply:
See #155